underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsPositionedUnderFileInputStream.java

```diff
@@ -61,22 +80,50 @@ public int read(byte[] b) throws IOException {
     return read(b, 0, b.length);
   }
 
+  /**
+   * If there has been a number of sequential reads in a row,
+   * we move to regular buffered reads.
+   */
   @Override
   public int read(byte[] buffer, int offset, int length) throws IOException {
-    int bytesRead = ((FSDataInputStream) in).read(mPos, buffer, offset, length);
+    int bytesRead;
+    if (mSequentialReadCount >= SEQUENTIAL_READ_LIMIT) {
+      ((Seekable) in).seek(mPos);
+      mSequentialReadCount = 0;
+    }
+    if (mPos == ((Seekable) in).getPos()) {
+      // same position, use buffered reads as default
+      bytesRead = in.read(buffer, offset, length);
+    } else {
+      bytesRead = ((PositionedReadable) in).read(mPos, buffer, offset, length);
+    }
     if (bytesRead > 0) {
       mPos += bytesRead;
+      mSequentialReadCount++;
     }
     return bytesRead;
   }
 
   @Override
   public void seek(long position) throws IOException {
+    if (position < mPos || position - mPos > MOVEMENT_LIMIT) {
+      mSequentialReadCount = 0;
+    } else {
+      ((Seekable) in).seek(position);
+    }
     mPos = position;
   }
 
   @Override
   public long skip(long n) throws IOException {
+    if (n <= 0) {
+      return 0;
+    }
+    if (n > MOVEMENT_LIMIT) {
+      mSequentialReadCount = 0;
+    } else {
+      ((Seekable) in).seek(mPos + n);
```

Comment:
nit: for readability i prefer a method called  `isSequentialReadMode` to determine whether to seek or not

Reply:
done